arthur/exts/directory/ldap.py

```diff
@@ -0,0 +1,416 @@
+
+        diff, missing_emp = await self.get_user_diff()
+
+        add_users = len([user for user in diff if user.action == LDAPSyncAction.ADD])
+        remove_users = len([user for user in diff if user.action == LDAPSyncAction.REMOVE])
+        keep_users = len([user for user in diff if user.action == LDAPSyncAction.KEEP])
+        change_users = len([user for user in diff if user.action == LDAPSyncAction.CHANGE])
+
+        logger.info(
+            f"LDAP: {add_users} missing users, removing {remove_users} users, "
+            f"keeping {keep_users} users, and changing {change_users} users."
+        )
+
+        if len(missing_emp) > 0:
+            logger.error(
+                "LDAP: Some users are missing an employee number. This may lead to duplicated users being created."
+            )
+
+            await self.bot.get_channel(CONFIG.devops_channel_id).send(
+                ":x: LDAP Sync: Some users are missing an employee number. This may lead to duplicate users, please rectify."
+            )
+
+        notified_users = []
+
+        async for message in self.bot.get_channel(CONFIG.ldap_bootstrap_channel_id).history(
+            limit=None, oldest_first=True
+        ):
+            if (
+                "Python Discord LDAP enrollment" in message.content
+                or len(message.mentions) == 0
+                or message.author != self.bot.user
+            ):
+                continue
+
+            notified_users.append(message.mentions[0])
+
+        for user in diff:
+            if user.action == LDAPSyncAction.ADD:
+                if user.discord_user in notified_users:
+                    continue
+
+                if NOTIFICATIONS_ENABLED:
+                    await self.bot.get_channel(CONFIG.ldap_bootstrap_channel_id).send(
+                        ELIGIBLE_MESSAGE.format(mention=user.discord_user.mention)
+                    )
+            if user.action == LDAPSyncAction.REMOVE:
+                freeipa.deactivate_user(user.ldap_user.uid)
+            elif user.action == LDAPSyncAction.CHANGE:
+                freeipa.set_user_groups(user.ldap_user.uid, user.groups)
+
+        logger.info("LDAP: Sync complete.")
+
+    async def clearup_bootstrap(self, user: discord.Member) -> None:
+        """Clear up the bootstrap message for a user."""
+        channel = self.bot.get_channel(CONFIG.ldap_bootstrap_channel_id)
+
+        async for message in channel.history(limit=None, oldest_first=True):
+            if message.author == self.bot.user and user.mention in message.content:
+                await message.delete()
+                break
+
+    @commands.Cog.listener()
+    async def on_member_update(self, before: discord.Member, after: discord.Member) -> None:
+        """Handle member updates."""
+        if before.roles == after.roles:
+            return
+
+        before_roles = set(role.id for role in before.roles)
+        after_roles = set(role.id for role in after.roles)
```

Comment:
Use a set comprehension instead
```suggestion
        before_roles = {role.id for role in before.roles}
        after_roles = {role.id for role in after.roles}
```

Reply:
Actually, it might be a bit faster to use genexprs here. Since you're only using `in` once. But maybe you love sets